File: implementations/java-springboot/harness/test/RuleTest.java
// Per-rule regression tests — run against testdata/<rule>/good|bad-*/.
//
// No external test framework such as JUnit is pulled in; this is a lightweight runner
// written with nothing but its own assert methods that throw AssertionError on failure
// (the same idea as the nestjs harness's run-fixtures.ts). It assumes it's run from the
// test/ directory and uses the "testdata/..." relative paths as-is.
import harness.Kind;
import harness.RuleResult;
import harness.rules.AggregateIdFormat;
import harness.rules.AggregateNoPublicSetters;
import harness.rules.ControllerPlacement;
import harness.rules.CqrsQueryPurity;
import harness.rules.DockerfileConventions;
import harness.rules.DomainLayerIsolation;
import harness.rules.DomainPurity;
import harness.rules.ErrorResponseSchema;
import harness.rules.EventPlacement;
import harness.rules.FileNaming;
import harness.rules.InterfaceNoInfrastructure;
import harness.rules.NoCrossAggregateReference;
import harness.rules.NoCrossBcDomainImport;
import harness.rules.NoCrossBcRepositoryInApplication;
import harness.rules.NoDirectEnvAccessOutsideConfig;
import harness.rules.NoEventPublisherInCommand;
import harness.rules.NoGenericResponseKeys;
import harness.rules.NoLoggingInDomain;
import harness.rules.NoOrmAutoSyncInProdConfig;
import harness.rules.NoSilentCatch;
import harness.rules.OutboxDrainOrder;
import harness.rules.PackageStructure;
import harness.rules.QueryHandlerNoRawAggregate;
import harness.rules.RateLimitWired;
import harness.rules.RepositoryAnnotation;
import harness.rules.RepositoryNaming;
import harness.rules.SchedulerInInfrastructureOnly;
import harness.rules.ServiceAnnotation;
import harness.rules.SharedInfra;
import harness.rules.SoftDeleteFilter;
import harness.rules.TransactionBoundary;
import harness.rules.TypedErrorsOnly;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

public final class RuleTest {
    private RuleTest() {
    }

    private static void assertNoFailures(RuleResult result) {
        long failures = result.count(Kind.FAIL);
        if (failures > 0) {
            throw new AssertionError("expected no failures, got: " + describeFailures(result));
        }
    }

    private static void assertHasFailure(RuleResult result) {
        if (result.count(Kind.FAIL) == 0) {
            throw new AssertionError("expected at least one failure, got none");
        }
    }

    private static String describeFailures(RuleResult result) {
        StringBuilder sb = new StringBuilder();
        for (var f : result.findings) {
            if (f.kind == Kind.FAIL) sb.append(f.name).append(" — ").append(f.reason).append("; ");
        }
        return sb.toString();
    }

    private record TestCase(String name, Runnable run) {
    }

    private static final List<TestCase> TESTS = List.of(
        new TestCase("file-naming/good", () -> assertNoFailures(FileNaming.check("testdata/file-naming/good"))),
        new TestCase("file-naming/bad-camelcase", () -> assertHasFailure(FileNaming.check("testdata/file-naming/bad-camelcase"))),

        new TestCase("repository-annotation/good", () -> assertNoFailures(RepositoryAnnotation.check("testdata/repository-annotation/good"))),
        new TestCase("repository-annotation/bad-outside-infrastructure", () -> assertHasFailure(RepositoryAnnotation.check("testdata/repository-annotation/bad-outside-infrastructure"))),

        new TestCase("service-annotation/good", () -> assertNoFailures(ServiceAnnotation.check("testdata/service-annotation/good"))),
        new TestCase("service-annotation/bad-outside-application", () -> assertHasFailure(ServiceAnnotation.check("testdata/service-annotation/bad-outside-application"))),

        new TestCase("domain-purity/good", () -> assertNoFailures(DomainPurity.check("testdata/domain-purity/good"))),
        new TestCase("domain-purity/bad-forbidden-annotation", () -> assertHasFailure(DomainPurity.check("testdata/domain-purity/bad-forbidden-annotation"))),

        new TestCase("controller-placement/good", () -> assertNoFailures(ControllerPlacement.check("testdata/controller-placement/good"))),
        new TestCase("controller-placement/bad-outside-interfaces", () -> assertHasFailure(ControllerPlacement.check("testdata/controller-placement/bad-outside-interfaces"))),
        new TestCase("controller-placement/good-advice-outside-interfaces", () -> assertNoFailures(ControllerPlacement.check("testdata/controller-placement/good-advice-outside-interfaces"))),

        new TestCase("package-structure/good", () -> assertNoFailures(PackageStructure.check("testdata/package-structure/good"))),
        new TestCase("package-structure/bad-missing-layer", () -> assertHasFailure(PackageStructure.check("testdata/package-structure/bad-missing-layer"))),

        new TestCase("shared-infra/good", () -> assertNoFailures(SharedInfra.check("testdata/shared-infra/good"))),
        new TestCase("shared-infra/bad-outbox-missing-files", () -> assertHasFailure(SharedInfra.check("testdata/shared-infra/bad-outbox-missing-files"))),
        new TestCase("shared-infra/bad-task-queue-misplaced", () -> assertHasFailure(SharedInfra.check("testdata/shared-infra/bad-task-queue-misplaced"))),

        new TestCase("event-placement/good", () -> assertNoFailures(EventPlacement.check("testdata/event-placement/good"))),
        new TestCase("event-placement/bad-wrong-dir", () -> assertHasFailure(EventPlacement.check("testdata/event-placement/bad-wrong-dir"))),
        new TestCase("event-placement/good-outbox-exempt", () -> assertNoFailures(EventPlacement.check("testdata/event-placement/good-outbox-exempt"))),
        new TestCase("event-placement/good-integration-event", () -> assertNoFailures(EventPlacement.check("testdata/event-placement/good-integration-event"))),
        new TestCase("event-placement/bad-integration-event-wrong-dir", () -> assertHasFailure(EventPlacement.check("testdata/event-placement/bad-integration-event-wrong-dir"))),

        new TestCase("no-event-publisher-in-command/good", () -> assertNoFailures(NoEventPublisherInCommand.check("testdata/no-event-publisher-in-command/good"))),
        new TestCase("no-event-publisher-in-command/bad-has-publisher", () -> assertHasFailure(NoEventPublisherInCommand.check("testdata/no-event-publisher-in-command/bad-has-publisher"))),

        new TestCase("transaction-boundary/good", () -> assertNoFailures(TransactionBoundary.check("testdata/transaction-boundary/good"))),
        new TestCase("transaction-boundary/bad-has-transactional", () -> assertHasFailure(TransactionBoundary.check("testdata/transaction-boundary/bad-has-transactional"))),
        new TestCase("transaction-boundary/bad-repository-impl-missing-transactional", () -> assertHasFailure(TransactionBoundary.check("testdata/transaction-boundary/bad-repository-impl-missing-transactional"))),

        new TestCase("outbox-drain-order/good", () -> assertNoFailures(OutboxDrainOrder.check("testdata/outbox-drain-order/good"))),
        new TestCase("outbox-drain-order/bad-references-outbox-relay", () -> assertHasFailure(OutboxDrainOrder.check("testdata/outbox-drain-order/bad-references-outbox-relay"))),
        new TestCase("outbox-drain-order/bad-calls-poller-directly", () -> assertHasFailure(OutboxDrainOrder.check("testdata/outbox-drain-order/bad-calls-poller-directly"))),

        new TestCase("cqrs-query-purity/good", () -> assertNoFailures(CqrsQueryPurity.check("testdata/cqrs-query-purity/good"))),
        new TestCase("cqrs-query-purity/bad-repository-reference", () -> assertHasFailure(CqrsQueryPurity.check("testdata/cqrs-query-purity/bad-repository-reference"))),

        new TestCase("repository-naming/good", () -> assertNoFailures(RepositoryNaming.check("testdata/repository-naming/good"))),
        new TestCase("repository-naming/bad-findby", () -> assertHasFailure(RepositoryNaming.check("testdata/repository-naming/bad-findby"))),
        new TestCase("repository-naming/bad-findall", () -> assertHasFailure(RepositoryNaming.check("testdata/repository-naming/bad-findall"))),
        new TestCase("repository-naming/bad-count", () -> assertHasFailure(RepositoryNaming.check("testdata/repository-naming/bad-count"))),
        new TestCase("repository-naming/bad-bare-save", () -> assertHasFailure(RepositoryNaming.check("testdata/repository-naming/bad-bare-save"))),
        new TestCase("repository-naming/bad-bare-delete", () -> assertHasFailure(RepositoryNaming.check("testdata/repository-naming/bad-bare-delete"))),
        new TestCase("repository-naming/bad-update", () -> assertHasFailure(RepositoryNaming.check("testdata/repository-naming/bad-update"))),

        new TestCase("domain-layer-isolation/good", () -> assertNoFailures(DomainLayerIsolation.check("testdata/domain-layer-isolation/good"))),
        new TestCase("domain-layer-isolation/bad-imports-application", () -> assertHasFailure(DomainLayerIsolation.check("testdata/domain-layer-isolation/bad-imports-application"))),
        new TestCase("domain-layer-isolation/bad-imports-sibling-infrastructure", () -> assertHasFailure(DomainLayerIsolation.check("testdata/domain-layer-isolation/bad-imports-sibling-infrastructure"))),

        new TestCase("interface-no-infrastructure/good", () -> assertNoFailures(InterfaceNoInfrastructure.check("testdata/interface-no-infrastructure/good"))),
        new TestCase("interface-no-infrastructure/bad-imports-infrastructure", () -> assertHasFailure(InterfaceNoInfrastructure.check("testdata/interface-no-infrastructure/bad-imports-infrastructure"))),

        new TestCase("aggregate-no-public-setters/good", () -> assertNoFailures(AggregateNoPublicSetters.check("testdata/aggregate-no-public-setters/good"))),
        new TestCase("aggregate-no-public-setters/bad-javabean-setter", () -> assertHasFailure(AggregateNoPublicSetters.check("testdata/aggregate-no-public-setters/bad-javabean-setter"))),
        new TestCase("aggregate-no-public-setters/bad-lombok-setter", () -> assertHasFailure(AggregateNoPublicSetters.check("testdata/aggregate-no-public-setters/bad-lombok-setter"))),

        new TestCase("no-cross-aggregate-reference/good", () -> assertNoFailures(NoCrossAggregateReference.check("testdata/no-cross-aggregate-reference/good"))),
        new TestCase("no-cross-aggregate-reference/bad-payment-references-refund", () -> assertHasFailure(NoCrossAggregateReference.check("testdata/no-cross-aggregate-reference/bad-payment-references-refund"))),

        new TestCase("no-direct-env-access-outside-config/good", () -> assertNoFailures(NoDirectEnvAccessOutsideConfig.check("testdata/no-direct-env-access-outside-config/good"))),
        new TestCase("no-direct-env-access-outside-config/bad-getenv-in-application", () -> assertHasFailure(NoDirectEnvAccessOutsideConfig.check("testdata/no-direct-env-access-outside-config/bad-getenv-in-application"))),

        new TestCase("no-cross-bc-repository-in-application/good", () -> assertNoFailures(NoCrossBcRepositoryInApplication.check("testdata/no-cross-bc-repository-in-application/good"))),
        new TestCase("no-cross-bc-repository-in-application/bad-cross-domain-repository", () -> assertHasFailure(NoCrossBcRepositoryInApplication.check("testdata/no-cross-bc-repository-in-application/bad-cross-domain-repository"))),
        new TestCase("no-cross-bc-repository-in-application/bad-cross-domain-query", () -> assertHasFailure(NoCrossBcRepositoryInApplication.check("testdata/no-cross-bc-repository-in-application/bad-cross-domain-query"))),

        new TestCase("no-logging-in-domain/good", () -> assertNoFailures(NoLoggingInDomain.check("testdata/no-logging-in-domain/good"))),
        new TestCase("no-logging-in-domain/bad-slf4j-in-domain", () -> assertHasFailure(NoLoggingInDomain.check("testdata/no-logging-in-domain/bad-slf4j-in-domain"))),

        new TestCase("scheduler-in-infrastructure-only/good", () -> assertNoFailures(SchedulerInInfrastructureOnly.check("testdata/scheduler-in-infrastructure-only/good"))),
        new TestCase("scheduler-in-infrastructure-only/bad-scheduled-in-application", () -> assertHasFailure(SchedulerInInfrastructureOnly.check("testdata/scheduler-in-infrastructure-only/bad-scheduled-in-application"))),

        new TestCase("no-silent-catch/good", () -> assertNoFailures(NoSilentCatch.check("testdata/no-silent-catch/good"))),
        new TestCase("no-silent-catch/bad-empty-catch", () -> assertHasFailure(NoSilentCatch.check("testdata/no-silent-catch/bad-empty-catch"))),

        new TestCase("dockerfile-conventions/good", () -> assertNoFailures(DockerfileConventions.check("testdata/dockerfile-conventions/good"))),
        new TestCase("dockerfile-conventions/bad-single-stage-no-healthcheck", () -> assertHasFailure(DockerfileConventions.check("testdata/dockerfile-conventions/bad-single-stage-no-healthcheck"))),
        new TestCase("dockerfile-conventions/bad-no-user", () -> assertHasFailure(DockerfileConventions.check("testdata/dockerfile-conventions/bad-no-user"))),

        new TestCase("aggregate-id-format/good", () -> assertNoFailures(AggregateIdFormat.check("testdata/aggregate-id-format/good"))),
        new TestCase("aggregate-id-format/bad-raw-uuid", () -> assertHasFailure(AggregateIdFormat.check("testdata/aggregate-id-format/bad-raw-uuid"))),

        new TestCase("error-response-schema/good", () -> assertNoFailures(ErrorResponseSchema.check("testdata/error-response-schema/good"))),
        new TestCase("error-response-schema/bad-missing-field", () -> assertHasFailure(ErrorResponseSchema.check("testdata/error-response-schema/bad-missing-field"))),
        new TestCase("error-response-schema/bad-extra-field", () -> assertHasFailure(ErrorResponseSchema.check("testdata/error-response-schema/bad-extra-field"))),

        new TestCase("soft-delete-filter/good", () -> assertNoFailures(SoftDeleteFilter.check("testdata/soft-delete-filter/good"))),
        new TestCase("soft-delete-filter/bad-missing-filter", () -> assertHasFailure(SoftDeleteFilter.check("testdata/soft-delete-filter/bad-missing-filter"))),
        new TestCase("soft-delete-filter/good-global-annotation", () -> assertNoFailures(SoftDeleteFilter.check("testdata/soft-delete-filter/good-global-annotation"))),

        new TestCase("typed-errors-only/good", () -> assertNoFailures(TypedErrorsOnly.check("testdata/typed-errors-only/good"))),
        new TestCase("typed-errors-only/bad-runtime-exception", () -> assertHasFailure(TypedErrorsOnly.check("testdata/typed-errors-only/bad-runtime-exception"))),
        new TestCase("typed-errors-only/bad-illegal-state", () -> assertHasFailure(TypedErrorsOnly.check("testdata/typed-errors-only/bad-illegal-state"))),

        new TestCase("rate-limit-wired/good", () -> assertNoFailures(RateLimitWired.check("testdata/rate-limit-wired/good"))),
        new TestCase("rate-limit-wired/bad-hardcoded", () -> assertHasFailure(RateLimitWired.check("testdata/rate-limit-wired/bad-hardcoded"))),
        new TestCase("rate-limit-wired/bad-not-component", () -> assertHasFailure(RateLimitWired.check("testdata/rate-limit-wired/bad-not-component"))),
        new TestCase("rate-limit-wired/bad-disabled", () -> assertHasFailure(RateLimitWired.check("testdata/rate-limit-wired/bad-disabled"))),

        new TestCase("no-generic-response-keys/good", () -> assertNoFailures(NoGenericResponseKeys.check("testdata/no-generic-response-keys/good"))),
        new TestCase("no-generic-response-keys/bad-data-field", () -> assertHasFailure(NoGenericResponseKeys.check("testdata/no-generic-response-keys/bad-data-field"))),
        new TestCase("no-generic-response-keys/bad-items-field", () -> assertHasFailure(NoGenericResponseKeys.check("testdata/no-generic-response-keys/bad-items-field"))),

        new TestCase("query-handler-no-raw-aggregate/good", () -> assertNoFailures(QueryHandlerNoRawAggregate.check("testdata/query-handler-no-raw-aggregate/good"))),
        new TestCase("query-handler-no-raw-aggregate/bad-service-returns-raw-aggregate", () -> assertHasFailure(QueryHandlerNoRawAggregate.check("testdata/query-handler-no-raw-aggregate/bad-service-returns-raw-aggregate"))),
        new TestCase("query-handler-no-raw-aggregate/bad-controller-returns-raw-aggregate", () -> assertHasFailure(QueryHandlerNoRawAggregate.check("testdata/query-handler-no-raw-aggregate/bad-controller-returns-raw-aggregate"))),

        new TestCase("no-cross-bc-domain-import/good", () -> assertNoFailures(NoCrossBcDomainImport.check("testdata/no-cross-bc-domain-import/good"))),
        new TestCase("no-cross-bc-domain-import/bad-card-imports-payment-domain", () -> assertHasFailure(NoCrossBcDomainImport.check("testdata/no-cross-bc-domain-import/bad-card-imports-payment-domain"))),

        new TestCase("no-orm-autosync-in-prod-config/good", () -> assertNoFailures(NoOrmAutoSyncInProdConfig.check("testdata/no-orm-autosync-in-prod-config/good"))),
        new TestCase("no-orm-autosync-in-prod-config/bad-prod-update", () -> assertHasFailure(NoOrmAutoSyncInProdConfig.check("testdata/no-orm-autosync-in-prod-config/bad-prod-update"))),
        new TestCase("no-orm-autosync-in-prod-config/bad-default-update", () -> assertHasFailure(NoOrmAutoSyncInProdConfig.check("testdata/no-orm-autosync-in-prod-config/bad-default-update")))
    );

    public static void main(String[] args) {
        int failures = 0;
        for (TestCase t : TESTS) {
            try {
                t.run().run();
                System.out.println("  PASS  " + t.name());
            } catch (AssertionError e) {
                failures++;
                System.out.println("  FAIL  " + t.name() + " — " + e.getMessage());
            }
        }
        System.out.println();
        if (failures == 0) {
            System.out.println(TESTS.size() + " passed  PASS");
        } else {
            System.out.println((TESTS.size() - failures) + " passed, " + failures + " failed  FAIL");
            System.exit(1);
        }
    }
}

File: implementations/kotlin-springboot/harness/test/RuleTest.kt
// 규칙별 회귀 테스트 — testdata/<rule>/good|bad-*/ 를 대상으로 실행한다.
//
// JUnit 등 외부 테스트 프레임워크를 새로 끌어오지 않고, 실패 시 AssertionError를 던지는
// 자체 assert 함수 하나만으로 작성된 가벼운 러너다(nestjs harness의 run-fixtures.ts와
// 같은 취지 — kotlin.test 아티팩트조차 별도 의존성이라 끌어오지 않는다). `test/`
// 디렉토리에서 실행한다고 가정하고 "testdata/..." 상대경로를 그대로 쓴다.
package harness.test

import harness.Kind
import harness.RuleResult
import harness.rules.checkAggregateNoPublicSetters
import harness.rules.checkControllerPlacement
import harness.rules.checkCqrsPattern
import harness.rules.checkDockerfileConventions
import harness.rules.checkDomainLayerIsolation
import harness.rules.checkDomainPurity
import harness.rules.checkEventPlacement
import harness.rules.checkFileNaming
import harness.rules.checkInterfaceNoInfrastructure
import harness.rules.checkNoCrossAggregateReference
import harness.rules.checkNoCrossBcRepositoryInApplication
import harness.rules.checkNoDirectEnvAccessOutsideConfig
import harness.rules.checkNoEventPublisherInCommand
import harness.rules.checkNoLoggingInDomain
import harness.rules.checkNoSilentCatch
import harness.rules.checkNotificationE2eTest
import harness.rules.checkOutboxNoSyncDrain
import harness.rules.checkPackageStructure
import harness.rules.checkRepositoryAnnotation
import harness.rules.checkRepositoryNaming
import harness.rules.checkSchedulerInInfrastructureOnly
import harness.rules.checkSealedException
import harness.rules.checkServiceAnnotation
import harness.rules.checkSharedInfra
import harness.rules.checkTransactionBoundary
import kotlin.system.exitProcess

private fun assertTrue(condition: Boolean, message: String) {
    if (!condition) throw AssertionError(message)
}

private fun RuleResult.assertNoFailures() {
    val failures = findings.filter { it.kind == Kind.FAIL }
    assertTrue(failures.isEmpty(), "expected no failures, got: ${failures.map { "${it.name} — ${it.reason}" }}")
}

private fun RuleResult.assertHasFailure() {
    assertTrue(findings.any { it.kind == Kind.FAIL }, "expected at least one failure, got none")
}

data class TestCase(val name: String, val run: () -> Unit)

val TESTS: List<TestCase> = listOf(
    TestCase("file-naming/good") { checkFileNaming("testdata/file-naming/good").assertNoFailures() },
    TestCase("file-naming/bad-camelcase") { checkFileNaming("testdata/file-naming/bad-camelcase").assertHasFailure() },

    TestCase("repository-annotation/good") { checkRepositoryAnnotation("testdata/repository-annotation/good").assertNoFailures() },
    TestCase("repository-annotation/bad-outside-infrastructure") { checkRepositoryAnnotation("testdata/repository-annotation/bad-outside-infrastructure").assertHasFailure() },

    TestCase("service-annotation/good") { checkServiceAnnotation("testdata/service-annotation/good").assertNoFailures() },
    TestCase("service-annotation/bad-outside-application") { checkServiceAnnotation("testdata/service-annotation/bad-outside-application").assertHasFailure() },

    TestCase("domain-purity/good") { checkDomainPurity("testdata/domain-purity/good").assertNoFailures() },
    TestCase("domain-purity/bad-forbidden-annotation") { checkDomainPurity("testdata/domain-purity/bad-forbidden-annotation").assertHasFailure() },
    TestCase("domain-purity/bad-jpa-annotation") { checkDomainPurity("testdata/domain-purity/bad-jpa-annotation").assertHasFailure() },

    TestCase("controller-placement/good") { checkControllerPlacement("testdata/controller-placement/good").assertNoFailures() },
    TestCase("controller-placement/bad-outside-interfaces") { checkControllerPlacement("testdata/controller-placement/bad-outside-interfaces").assertHasFailure() },
    TestCase("controller-placement/good-restcontrolleradvice-outside-interfaces") { checkControllerPlacement("testdata/controller-placement/good-restcontrolleradvice-outside-interfaces").assertNoFailures() },

    TestCase("sealed-exception/good") { checkSealedException("testdata/sealed-exception/good").assertNoFailures() },
    TestCase("sealed-exception/bad-outside-domain") { checkSealedException("testdata/sealed-exception/bad-outside-domain").assertHasFailure() },
    TestCase("sealed-exception/good-comment-mentions-sealed-class") { checkSealedException("testdata/sealed-exception/good-comment-mentions-sealed-class").assertNoFailures() },

    TestCase("package-structure/good") { checkPackageStructure("testdata/package-structure/good").assertNoFailures() },
    TestCase("package-structure/bad-missing-layer") { checkPackageStructure("testdata/package-structure/bad-missing-layer").assertHasFailure() },

    TestCase("shared-infra/good") { checkSharedInfra("testdata/shared-infra/good").assertNoFailures() },
    TestCase("shared-infra/bad-outbox-missing-files") { checkSharedInfra("testdata/shared-infra/bad-outbox-missing-files").assertHasFailure() },
    TestCase("shared-infra/bad-task-queue-misplaced") { checkSharedInfra("testdata/shared-infra/bad-task-queue-misplaced").assertHasFailure() },

    TestCase("event-placement/good") { checkEventPlacement("testdata/event-placement/good").assertNoFailures() },
    TestCase("event-placement/bad-wrong-dir") { checkEventPlacement("testdata/event-placement/bad-wrong-dir").assertHasFailure() },

    TestCase("no-event-publisher-in-command/good") { checkNoEventPublisherInCommand("testdata/no-event-publisher-in-command/good").assertNoFailures() },
    TestCase("no-event-publisher-in-command/bad-has-publisher") { checkNoEventPublisherInCommand("testdata/no-event-publisher-in-command/bad-has-publisher").assertHasFailure() },

    TestCase("transaction-boundary/good") { checkTransactionBoundary("testdata/transaction-boundary/good").assertNoFailures() },
    TestCase("transaction-boundary/bad-has-transactional") { checkTransactionBoundary("testdata/transaction-boundary/bad-has-transactional").assertHasFailure() },
    TestCase("transaction-boundary/bad-repository-impl-missing-transactional") { checkTransactionBoundary("testdata/transaction-boundary/bad-repository-impl-missing-transactional").assertHasFailure() },

    TestCase("outbox-no-sync-drain/good") { checkOutboxNoSyncDrain("testdata/outbox-no-sync-drain/good").assertNoFailures() },
    TestCase("outbox-no-sync-drain/bad-references-relay") { checkOutboxNoSyncDrain("testdata/outbox-no-sync-drain/bad-references-relay").assertHasFailure() },
    TestCase("outbox-no-sync-drain/bad-calls-process-pending") { checkOutboxNoSyncDrain("testdata/outbox-no-sync-drain/bad-calls-process-pending").assertHasFailure() },

    TestCase("notification-e2e-test/good") { checkNotificationE2eTest("testdata/notification-e2e-test/good").assertNoFailures() },
    TestCase("notification-e2e-test/bad-missing") { checkNotificationE2eTest("testdata/notification-e2e-test/bad-missing").assertHasFailure() },

    TestCase("cqrs-pattern/good") { checkCqrsPattern("testdata/cqrs-pattern/good").assertNoFailures() },
    TestCase("cqrs-pattern/bad-query-uses-repository") { checkCqrsPattern("testdata/cqrs-pattern/bad-query-uses-repository").assertHasFailure() },

    TestCase("repository-naming/good") { checkRepositoryNaming("testdata/repository-naming/good").assertNoFailures() },
    TestCase("repository-naming/good-infra-excluded") { checkRepositoryNaming("testdata/repository-naming/good-infra-excluded").assertNoFailures() },
    TestCase("repository-naming/bad-findby") { checkRepositoryNaming("testdata/repository-naming/bad-findby").assertHasFailure() },
    TestCase("repository-naming/bad-findall") { checkRepositoryNaming("testdata/repository-naming/bad-findall").assertHasFailure() },
    TestCase("repository-naming/bad-count") { checkRepositoryNaming("testdata/repository-naming/bad-count").assertHasFailure() },
    TestCase("repository-naming/bad-bare-save") { checkRepositoryNaming("testdata/repository-naming/bad-bare-save").assertHasFailure() },
    TestCase("repository-naming/bad-bare-delete") { checkRepositoryNaming("testdata/repository-naming/bad-bare-delete").assertHasFailure() },
    TestCase("repository-naming/bad-update") { checkRepositoryNaming("testdata/repository-naming/bad-update").assertHasFailure() },

    TestCase("domain-layer-isolation/good") { checkDomainLayerIsolation("testdata/domain-layer-isolation/good").assertNoFailures() },
    TestCase("domain-layer-isolation/bad-same-domain-application") { checkDomainLayerIsolation("testdata/domain-layer-isolation/bad-same-domain-application").assertHasFailure() },
    TestCase("domain-layer-isolation/bad-sibling-domain-infrastructure") { checkDomainLayerIsolation("testdata/domain-layer-isolation/bad-sibling-domain-infrastructure").assertHasFailure() },

    TestCase("interface-no-infrastructure/good") { checkInterfaceNoInfrastructure("testdata/interface-no-infrastructure/good").assertNoFailures() },
    TestCase("interface-no-infrastructure/bad-imports-infrastructure") { checkInterfaceNoInfrastructure("testdata/interface-no-infrastructure/bad-imports-infrastructure").assertHasFailure() },

    TestCase("aggregate-no-public-setters/good") { checkAggregateNoPublicSetters("testdata/aggregate-no-public-setters/good").assertNoFailures() },
    TestCase("aggregate-no-public-setters/bad-public-var") { checkAggregateNoPublicSetters("testdata/aggregate-no-public-setters/bad-public-var").assertHasFailure() },

    TestCase("no-cross-aggregate-reference/good") { checkNoCrossAggregateReference("testdata/no-cross-aggregate-reference/good").assertNoFailures() },
    TestCase("no-cross-aggregate-reference/bad-payment-holds-refund") { checkNoCrossAggregateReference("testdata/no-cross-aggregate-reference/bad-payment-holds-refund").assertHasFailure() },
    TestCase("no-cross-aggregate-reference/bad-refund-holds-payment") { checkNoCrossAggregateReference("testdata/no-cross-aggregate-reference/bad-refund-holds-payment").assertHasFailure() },

    TestCase("no-direct-env-access-outside-config/good") { checkNoDirectEnvAccessOutsideConfig("testdata/no-direct-env-access-outside-config/good").assertNoFailures() },
    TestCase("no-direct-env-access-outside-config/bad-getenv-in-application") { checkNoDirectEnvAccessOutsideConfig("testdata/no-direct-env-access-outside-config/bad-getenv-in-application").assertHasFailure() },

    TestCase("no-cross-bc-repository-in-application/good") { checkNoCrossBcRepositoryInApplication("testdata/no-cross-bc-repository-in-application/good").assertNoFailures() },
    TestCase("no-cross-bc-repository-in-application/bad-cross-domain-repository") { checkNoCrossBcRepositoryInApplication("testdata/no-cross-bc-repository-in-application/bad-cross-domain-repository").assertHasFailure() },

    TestCase("no-logging-in-domain/good") { checkNoLoggingInDomain("testdata/no-logging-in-domain/good").assertNoFailures() },
    TestCase("no-logging-in-domain/bad-slf4j-in-domain") { checkNoLoggingInDomain("testdata/no-logging-in-domain/bad-slf4j-in-domain").assertHasFailure() },

    TestCase("scheduler-in-infrastructure-only/good") { checkSchedulerInInfrastructureOnly("testdata/scheduler-in-infrastructure-only/good").assertNoFailures() },
    TestCase("scheduler-in-infrastructure-only/bad-scheduled-in-application") { checkSchedulerInInfrastructureOnly("testdata/scheduler-in-infrastructure-only/bad-scheduled-in-application").assertHasFailure() },

    TestCase("no-silent-catch/good") { checkNoSilentCatch("testdata/no-silent-catch/good").assertNoFailures() },
    TestCase("no-silent-catch/bad-empty-catch") { checkNoSilentCatch("testdata/no-silent-catch/bad-empty-catch").assertHasFailure() },

    TestCase("dockerfile-conventions/good") { checkDockerfileConventions("testdata/dockerfile-conventions/good").assertNoFailures() },
    TestCase("dockerfile-conventions/bad-single-stage-no-healthcheck") { checkDockerfileConventions("testdata/dockerfile-conventions/bad-single-stage-no-healthcheck").assertHasFailure() },
    TestCase("dockerfile-conventions/bad-missing-dockerignore") { checkDockerfileConventions("testdata/dockerfile-conventions/bad-missing-dockerignore").assertHasFailure() }
)

fun main() {
    var failures = 0
    for (t in TESTS) {
        try {
            t.run()
            println("  PASS  ${t.name}")
        } catch (e: AssertionError) {
            failures++
            println("  FAIL  ${t.name} — ${e.message}")
        }
    }
    println()
    if (failures == 0) {
        println("${TESTS.size} passed  PASS")
    } else {
        println("${TESTS.size - failures} passed, $failures failed  FAIL")
        exitProcess(1)
    }
}
